src/Workspaces/Core/Portable/Utilities/ValuesSources/OptionalValueSource.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A class that abstracts the accessing of a value that might not be available anymore when queried.
+    /// </summary>
+    internal abstract class OptionalValueSource<T> where T : class
+    {
+        public abstract bool TryGetValue([NotNullWhen(true)]out T? value);
+        public abstract T? GetValue(CancellationToken cancellationToken = default);
+        public abstract Task<T?> GetValueAsync(CancellationToken cancellationToken = default);
+
+        public bool HasValue => TryGetValue(out _);
```

Comment:
💡 Consider marking this obsolete. There is no guarantee that a subsequent call to `GetValue` will return true, making this property useless. The alternative is making it virtual and always return `false` here, and updating the contract to state that it only returns `true` if the value is cached in a manner that it will definitely not recompute a value if `GetValue` is called.

Reply:
Inlined the method into the two places that used it.